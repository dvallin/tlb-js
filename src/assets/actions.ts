import { Action, Movement, Status, Attack } from '../components/action'
import { Effect, damage, confuse, stun, immobilize, defend, kill, heal, negateEffects } from '../components/effects'

function movement(range: number): Movement {
  return { kind: 'movement', range, target: 'self' }
}

function attack(range: number, accuracy: number, effects: Effect[]): Attack {
  return { kind: 'attack', range, accuracy, effects, target: 'enemy' }
}

function status(effects: Effect[]): Status {
  return { kind: 'status', effects, target: 'self' }
}

const actionsDefinition = {
  endTurn: {
    name: 'end turn',
    cost: {
      actions: 0,
      movement: 0,
      costsAll: true,
    },
    subActions: [],
  },
  shortMove: {
    name: 'move',
    cost: {
      actions: 0,
      movement: 2,
    },
    subActions: [movement(3)],
  },
  longMove: {
    name: 'move',
    cost: {
      actions: 0,
      movement: 2,
    },
    subActions: [movement(5)],
  },
  healLimp: {
    name: 'healLimp',
    cost: {
      actions: 1,
      movement: 0,
    },
    subActions: [status([heal(), negateEffects()])],
  },
  kill: {
    name: 'kill',
    cost: {
      actions: 1,
      movement: 0,
    },
    subActions: [status([kill()])],
  },
  hit: {
    name: 'hit',
    cost: {
      actions: 2,
      movement: 0,
    },
    subActions: [attack(1, 8, [damage(3)])],
  },
  strideAndSlip: {
    name: 'stride and slip',
    cost: {
      actions: 0,
      movement: 3,
    },
    subActions: [movement(8), status([immobilize(1)])],
  },
  rush: {
    name: 'rush',
    cost: {
      actions: 2,
      movement: 3,
      costsAll: true,
    },
    subActions: [movement(8)],
  },
  shoot: {
    name: 'shoot',
    cost: {
      actions: 3,
      movement: 0,
    },
    subActions: [attack(10, 4, [damage(5)])],
  },
  headshot: {
    name: 'headshot',
    cost: {
      actions: 4,
      movement: 0,
    },
    subActions: [attack(20, 10, [damage(5, ['head'])])],
  },
  overcharge: {
    name: 'overcharge',
    cost: {
      actions: 3,
      movement: 0,
    },
    attack: {},
    subActions: [attack(8, 4, [damage(3), immobilize(1)])],
  },
  bolt: {
    name: 'bolt',
    cost: {
      actions: 3,
      movement: 0,
    },
    subActions: [attack(1, 5, [damage(2, ['head']), stun(1)])],
  },
  hitAndRun: {
    name: 'hit and run',
    cost: {
      actions: 3,
      movement: 3,
    },
    subActions: [attack(1, 4, [damage(3), confuse(1)]), movement(6)],
  },
  tighten: {
    name: 'tighten',
    cost: {
      actions: 1,
      movement: 0,
    },
    subActions: [status([defend(2, 1)])],
  },
}
export type ActionType = keyof typeof actionsDefinition
export const actions: { [key in ActionType]: Action } = actionsDefinition
