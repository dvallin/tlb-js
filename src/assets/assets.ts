import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
