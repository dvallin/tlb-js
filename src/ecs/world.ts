import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
