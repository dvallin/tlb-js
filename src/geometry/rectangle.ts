import { AbstractShape } from "./shape"
import { Vector } from "../spatial"

export class Rectangle extends AbstractShape {

    public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
        return new Rectangle(left, top, right - left + 1, bottom - top + 1)
    }

    public constructor(
        public readonly x: number,
        public readonly y: number,
        public readonly width: number,
        public readonly height: number,
    ) {
        super()
    }

    public get left(): number {
        return this.x
    }

    public get right(): number {
        return this.x + this.width - 1
    }

    public get top(): number {
        return this.y
    }

    public get bottom(): number {
        return this.y + this.height - 1
    }

    public get topLeft(): Vector {
        return new Vector(this.left, this.top)
    }

    public get topRight(): Vector {
        return new Vector(this.right, this.top)
    }

    public get bottomRight(): Vector {
        return new Vector(this.right, this.bottom)
    }

    public get bottomLeft(): Vector {
        return new Vector(this.left, this.bottom)
    }

    public get center(): Vector {
        const cx = (this.right + this.left) / 2
        const cy = (this.bottom + this.top) / 2
        return new Vector(Math.floor(cx), Math.floor(cy))
    }

    public plus(other: Rectangle): Rectangle {
        return Rectangle.fromBounds(
            Math.min(this.left, other.left), Math.max(this.right, other.right),
            Math.min(this.top, other.top), Math.max(this.bottom, other.bottom)
        )
    }

    public bounds(): Rectangle {
        return this
    }

    public containsVector(p: Vector): boolean {
        return p.x >= this.x && p.x < this.x + this.width
            && p.y >= this.y && p.y < this.y + this.height
    }

    public translate(t: Vector): Rectangle {
        return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
    }

    public grow(): Rectangle {
        return new Rectangle(this.x - 1, this.y - 1, this.width + 2, this.height + 2)
    }

    public all(f: (position: Vector) => boolean): boolean {
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                const position = new Vector(j + this.x, i + this.y)
                if (!f(position)) {
                    return false
                }
            }
        }
        return true
    }
}
