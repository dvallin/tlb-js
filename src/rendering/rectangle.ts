import { Rectangle } from "@/geometry/Rectangle"
import { Position } from "@/geometry/Position"
import { RenderIterator } from "@/rendering"

export function rasterize(rectangle: Rectangle, fill: boolean = false): RenderIterator {
    let y = rectangle.top
    let x = rectangle.left
    let done = false

    return () => {
        if (done) {
            return undefined
        }
        const result = new Position(Math.floor(x), Math.floor(y))
        x++
        if (x > rectangle.right) {
            x = rectangle.left
            y++
        }
        if (!fill && x > rectangle.left && y !== rectangle.top && y !== rectangle.bottom) {
            x = rectangle.right
        }
        if (y > rectangle.bottom) {
            done = true
        }
        return result
    }
}
