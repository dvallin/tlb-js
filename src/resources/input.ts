import { TlbResource, ResourceName } from '../tlb'
import { Vector } from '../spatial'
import { Position } from '../renderer/position'
import { KEYS } from 'rot-js'

export interface Input {
  position: Position | undefined
  mouseDown: boolean
  mousePressed: boolean
  mouseReleased: boolean

  keyDown: Set<number>
  keyPressed: Set<number>
  keyReleased: Set<number>

  createMovementDelta(): Vector
}

export class InputResource implements TlbResource, Input {
  public readonly kind: ResourceName = 'input'

  public position: Position | undefined = undefined

  public mouseDown: boolean = false
  public mousePressed: boolean = false
  public mouseReleased: boolean = false

  public keyDown: Set<number> = new Set()
  public keyPressed: Set<number> = new Set()
  public keyReleased: Set<number> = new Set()

  private mouseEvent: MouseEvent | undefined = undefined
  private keyEvents: KeyboardEvent[] = []

  public constructor(public readonly eventToPosition: (event: MouseEvent | TouchEvent) => Position | undefined) {
    document.addEventListener('mousedown', e => (this.mouseEvent = e))
    document.addEventListener('mousemove', e => (this.mouseEvent = e))
    document.addEventListener('keydown', e => this.keyEvents.push(e))
    document.addEventListener('keyup', e => this.keyEvents.push(e))
  }

  public update(): void {
    this.handleMouseEvent()
    this.handleKeyboardEvents()
  }

  public createMovementDelta(): Vector {
    let delta = new Vector(0, 0)
    if (this.keyDown.has(KEYS.VK_H)) {
      delta = delta.add(Vector.fromDirection('left'))
    }
    if (this.keyDown.has(KEYS.VK_J)) {
      delta = delta.add(Vector.fromDirection('down'))
    }
    if (this.keyDown.has(KEYS.VK_K)) {
      delta = delta.add(Vector.fromDirection('up'))
    }
    if (this.keyDown.has(KEYS.VK_L)) {
      delta = delta.add(Vector.fromDirection('right'))
    }
    return delta.normalize()
  }

  private handleMouseEvent(): void {
    this.mouseReleased = false
    this.mousePressed = false
    if (this.mouseEvent) {
      this.position = this.eventToPosition(this.mouseEvent)
      const pressed = this.mouseEvent.buttons > 0
      if (pressed) {
        if (this.position) {
          this.mousePressed = true
          this.mouseDown = true
        }
      } else {
        this.mouseDown = false
        this.mouseReleased = true
      }
      this.mouseEvent = undefined
    }
  }

  private handleKeyboardEvents(): void {
    this.keyPressed.clear()
    this.keyReleased.clear()
    for (const e of this.keyEvents) {
      if (e.type === 'keydown') {
        this.keyPressed.add(e.keyCode)
        this.keyDown.add(e.keyCode)
      } else {
        this.keyReleased.add(e.keyCode)
        this.keyDown.delete(e.keyCode)
      }
    }
    this.keyEvents = []
  }
}
