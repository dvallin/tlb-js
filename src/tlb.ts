import { World } from './ecs/world'
import { VectorStorage, MapStorage, SetStorage, SingletonStorage } from './ecs/storage'
import { System } from './ecs/system'
import { Resource } from './ecs/resource'

import { AgeComponent } from './components/age'
import { AssetComponent } from './components/asset'
import { CharacterStatsComponent } from './components/character-stats'
import { FeatureComponent } from './components/feature'
import { FovComponent } from './components/fov'
import { GroundComponent } from './components/ground'
import { PositionComponent } from './components/position'
import { RegionComponent, StructureComponent } from './components/region'

import { Fov } from './systems/fov'
import { FreeModeControl } from './systems/free-mode-control'
import { Npc } from './systems/npc'
import { PlayerControl } from './systems/player-control'
import { PlayerInteraction } from './systems/player-interaction'
import { RegionCreator } from './systems/region-creator'
import { PlayerRoundControl } from './systems/player-round-control'
import { Trigger } from './systems/trigger'
import { ApplyEffects } from './systems/apply-effect'
import { Script } from './systems/script'

import { WorldMapResource } from './resources/world-map'
import { ViewportResource } from './resources/viewport'
import { InputResource } from './resources/input'
import { UIResource } from './resources/ui'
import { LogResource } from './resources/log'

import { Random } from './random'
import { Uniform } from './random/distributions'

import { Renderer } from './renderer/renderer'
import { Queries } from './renderer/queries'
import { State } from './game-states/state'
import { OverlayComponent } from './components/overlay'
import { TakeTurnComponent } from './components/rounds'
import { SelectedActionComponent, HasActionComponent } from './components/action'
import { ScriptComponent } from './components/script'
import { AiComponent } from './components/ai'
import { AiRoundControl } from './systems/ai-round-control'
import { TriggersComponent, TriggeredByComponent } from './components/trigger'
import { EffectComponent, ActiveEffectsComponent } from './components/effects'
import { InventoryComponent, ItemComponent, EquipedItemsComponent } from './components/items'
import { StartRound } from './systems/start-round'
import { DialogComponent } from './components/dialog'
export type ComponentName =
  | 'active'
  | 'age'
  | 'ai'
  | 'asset'
  | 'character-stats'
  | 'dead'
  | 'dialog'
  | 'effect'
  | 'equiped-items'
  | 'feature'
  | 'fov'
  | 'free-mode-anchor'
  | 'ground'
  | 'has-action'
  | 'inventory'
  | 'item'
  | 'npc'
  | 'overlay'
  | 'player'
  | 'position'
  | 'region'
  | 'structure'
  | 'script'
  | 'selected-action'
  | 'active-effects'
  | 'spawn'
  | 'start-turn'
  | 'take-turn'
  | 'took-turn'
  | 'triggered-by'
  | 'triggers'
  | 'viewport-focus'
  | 'wait-turn'
export type SystemName =
  | 'ai-round-control'
  | 'start-round'
  | 'effect'
  | 'fov'
  | 'free-mode-control'
  | 'info-popup'
  | 'npc'
  | 'player-control'
  | 'player-interaction'
  | 'player-round-control'
  | 'region-creator'
  | 'script'
  | 'trigger'
export type ResourceName = 'input' | 'map' | 'viewport' | 'ui' | 'log'

export type TlbWorld = World<ComponentName, SystemName, ResourceName>
export type TlbResource = Resource<ComponentName, SystemName, ResourceName>
export type TlbSystem = System<ComponentName, SystemName, ResourceName>

export function registerComponents<S, R>(world: World<ComponentName, S, R>): void {
  world.registerComponentStorage('active', new SetStorage())
  world.registerComponentStorage('age', new MapStorage<AgeComponent>())
  world.registerComponentStorage('ai', new MapStorage<AiComponent>())
  world.registerComponentStorage('asset', new MapStorage<AssetComponent>())
  world.registerComponentStorage('character-stats', new MapStorage<CharacterStatsComponent>())
  world.registerComponentStorage('dead', new SetStorage())
  world.registerComponentStorage('dialog', new MapStorage<DialogComponent>())
  world.registerComponentStorage('effect', new MapStorage<EffectComponent>())
  world.registerComponentStorage('feature', new VectorStorage<FeatureComponent>())
  world.registerComponentStorage('fov', new MapStorage<FovComponent>())
  world.registerComponentStorage('free-mode-anchor', new SingletonStorage<{}>())
  world.registerComponentStorage('ground', new MapStorage<GroundComponent>())
  world.registerComponentStorage('has-action', new MapStorage<HasActionComponent>())
  world.registerComponentStorage('inventory', new MapStorage<InventoryComponent>())
  world.registerComponentStorage('structure', new MapStorage<StructureComponent>())
  world.registerComponentStorage('equiped-items', new MapStorage<EquipedItemsComponent>())
  world.registerComponentStorage('item', new MapStorage<ItemComponent>())
  world.registerComponentStorage('npc', new SetStorage())
  world.registerComponentStorage('overlay', new MapStorage<OverlayComponent>())
  world.registerComponentStorage('player', new SingletonStorage<{}>())
  world.registerComponentStorage('position', new VectorStorage<PositionComponent>())
  world.registerComponentStorage('region', new MapStorage<RegionComponent>())
  world.registerComponentStorage('script', new MapStorage<ScriptComponent>())
  world.registerComponentStorage('selected-action', new SingletonStorage<SelectedActionComponent>())
  world.registerComponentStorage('active-effects', new MapStorage<ActiveEffectsComponent>())
  world.registerComponentStorage('spawn', new SingletonStorage<{}>())
  world.registerComponentStorage('take-turn', new MapStorage<TakeTurnComponent>())
  world.registerComponentStorage('took-turn', new SetStorage())
  world.registerComponentStorage('start-turn', new SetStorage())
  world.registerComponentStorage('triggered-by', new MapStorage<TriggeredByComponent>())
  world.registerComponentStorage('triggers', new MapStorage<TriggersComponent>())
  world.registerComponentStorage('viewport-focus', new SingletonStorage<{}>())
  world.registerComponentStorage('wait-turn', new SetStorage())
}

export function registerResources(
  world: World<ComponentName, SystemName, ResourceName>,
  renderer: Renderer,
  worldWidth: number = 256
): void {
  world.registerResource(new WorldMapResource(worldWidth))
  world.registerResource(new ViewportResource(renderer.boundaries))
  world.registerResource(new InputResource(e => renderer.eventToPosition(e)))
  world.registerResource(new UIResource())
  world.registerResource(new LogResource())
}

export function registerSystems(
  world: World<ComponentName, SystemName, ResourceName>,
  queries: Queries,
  pushState: (s: State) => void
): void {
  const uniform = new Uniform('some seed')
  world.registerSystem('ai-round-control', new AiRoundControl(queries, new Random(uniform)))
  world.registerSystem('effect', new ApplyEffects())
  world.registerSystem('fov', new Fov(queries))
  world.registerSystem('free-mode-control', new FreeModeControl())
  world.registerSystem('npc', new Npc(queries, new Random(uniform), pushState))
  world.registerSystem('player-control', new PlayerControl())
  world.registerSystem('player-interaction', new PlayerInteraction(pushState))
  world.registerSystem('player-round-control', new PlayerRoundControl(queries, new Random(uniform)))
  world.registerSystem('region-creator', new RegionCreator(uniform))
  world.registerSystem('script', new Script())
  world.registerSystem('trigger', new Trigger(new Random(uniform), pushState))
  world.registerSystem('start-round', new StartRound(new Random(uniform)))
}
