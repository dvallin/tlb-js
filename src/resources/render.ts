import { TlbResource, TlbWorld, ResourceName } from "../tlb"
import { getFeature } from "../components/feature"
import { PositionComponent } from "../components/position"
import { Renderer } from "../renderer/renderer"
import { Position } from "../renderer/position"
import { Viewport } from "./viewport"

export class Render implements TlbResource {

    public readonly kind: ResourceName = "render"

    public constructor(
        private readonly renderer: Renderer
    ) { }

    public update(world: TlbWorld): void {
        this.renderer.clear()
        const viewport = world.getResource<Viewport>("viewport")
        world.getStorage("in-viewport").foreach(entity => {
            const feature = getFeature(world, entity)
            const position = world.getComponent<PositionComponent>(entity, "position")
            if (feature && position) {
                const displayPosition = viewport.toDisplay(position.position)
                this.renderer.character(feature.character, displayPosition, feature.diffuse)
            }
        })
    }

    public eventToPosition(e: UIEvent): Position | undefined {
        return this.renderer.eventToPosition(e)
    }
}
