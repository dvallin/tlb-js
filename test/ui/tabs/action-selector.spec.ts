import { ActionSelectorFullView, ActionGroup } from '../../../src/ui/tabs/action-selector'
import { Rectangle } from '../../../src/geometry/rectangle'
import { actions } from '../../../src/assets/actions'
import { displayerRenderer } from '../../render'
import { World } from '../../../src/ecs/world'
import { TlbWorld } from '../../../src/tlb'
import { mockInput, mockImplementation } from '../../mocks'
import { KeyboardCommand, Input } from '../../../src/resources/input'

describe('ActionSelectorFullView', () => {
  let actionGroups: ActionGroup[]
  let view: ActionSelectorFullView
  let world: TlbWorld
  let input: Input
  beforeEach(() => {
    actionGroups = [actionGroup(1, true), actionGroup(2, false)]
    view = new ActionSelectorFullView(new Rectangle(0, 0, 20, 20), actionGroups)
    world = new World()
    input = mockInput(world)
  })

  it('displays action groups', () => {
    const { renderer, getDisplay } = displayerRenderer()
    view.render(renderer)
    expect(getDisplay()).toMatchSnapshot()
  })

  it('displays actions', () => {
    mockImplementation<KeyboardCommand, boolean>(input.isActive, key => key === 'accept')
    const { renderer, getDisplay } = displayerRenderer()
    view.update(world)
    view.render(renderer)
    expect(getDisplay()).toMatchSnapshot()
  })

  it('counts', () => {
    expect(view.length).toEqual(2)
  })

  it('hovers actions', () => {
    mockImplementation<KeyboardCommand, boolean>(input.isActive, key => key === 'accept')
    view.update(world)

    expect(view.hovered).toEqual({ action: actionGroups[0].items[0].action, entity: 1 })
  })

  it('selects actions', () => {
    mockImplementation<KeyboardCommand, boolean>(input.isActive, key => key === 'accept')

    view.update(world)
    view.update(world)

    expect(view.selected).toEqual({ action: actionGroups[0].items[0].action, entity: 1 })
  })
})

function actionGroup(id: number, available: boolean): ActionGroup {
  return {
    entity: id,
    name: `name${id}`,
    description: `desc${id}`,
    items: [
      {
        action: actions.hit,
        available,
      },
    ],
  }
}
