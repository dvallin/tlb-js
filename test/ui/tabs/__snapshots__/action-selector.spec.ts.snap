// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSelectorFullView displays action groups 1`] = `
"1 name1  name1      
2 name2  desc1      "
`;

exports[`ActionSelectorFullView displays actions 1`] = `
"1 hit    hit        

         cost: 2AP 0MP
          attack 3  "
`;
